File: framework/ruby/src/soda/unittest/convert.rb
require_relative '../leetcode/list'
require_relative '../leetcode/nest'
require_relative '../leetcode/tree'

module Soda end

module Soda::Unittest

class ObjectConverter
  def initialize(parser, serializer)
    @parser = parser
    @serializer = serializer
  end

  def from_json_serializable(j)
    @parser.call(j)
  end

  def to_json_serializable(obj)
    @serializer.call(obj)
  end
end

class ConverterFactory
  def self._register(type, parser, serializer)
    @factory_map[type.to_s] = ->() {
      ObjectConverter.new(parser, serializer)
    }
  end

  def self._init
    lc = Soda::Leetcode
    _register('ListNode', lc::ListFactory.method(:create), lc::ListFactory.method(:dump))
    _register(
      'ListNode[]',
      ->(ls) { ls.map { |e| lc::ListFactory.create(e) } },
      ->(ts) { ts.map { |e| lc::ListFactory.dump(e) } }
    )
    _register(
      'NestedInteger',
      lc::NestedIntegerFactory.method(:parse),
      lc::NestedIntegerFactory.method(:serialize)
    )
    _register(
      'NestedInteger[]',
      lc::NestedIntegerFactory.method(:parse_multi),
      lc::NestedIntegerFactory.method(:serialize_multi)
    )
    _register('TreeNode', lc::TreeFactory.method(:create), lc::TreeFactory.method(:dump))
  end

  def self.create(obj_type)
    @factory_map[obj_type].call
  end

  @factory_map = Hash.new(-> () {
    ObjectConverter.new(->(j){j}, ->(o){o})
  })
  _init
end

end  # module Soda::Unittest

File: framework/ts/src/soda/unittest/convert.ts
import {ListNode, ListFactory as LinkedListFactory} from '../leetcode/list';
import {NestedInteger, NestedIntegerFactory} from "../leetcode/nest";

class InternalConverter {

    constructor(private parser: (js: any) => any, private serializer: (obj: any) => any) {
    }

    fromJsonSerializer(js: any): any {
        return this.parser(js);
    }

    toJsonSerializer(obj: any): any {
        return this.serializer(obj);
    }

}

export class ObjectConverter<T> {

    constructor(private converter: InternalConverter) {
    }

    fromJsonSerializable(js: any): T {
        return this.converter.fromJsonSerializer(js) as T;
    }

    toJsonSerializable(obj: T): any {
        return this.converter.toJsonSerializer(obj);
    }

    internal(): InternalConverter {
        return this.converter;
    }
}

type FactoryType = (typ: string) => InternalConverter;
type ParserFunc<T> = (js: any) => T;
type SerialFunc<T> = (obj: T) => any;

export class ConverterFactory {

    private static factoryMap = new Map<string, FactoryType>();

    static create<T>(typeName: string): ObjectConverter<T> {
        let fact = this.factoryMap.get(typeName);
        let ic = fact ? fact(typeName) : new InternalConverter(j => j, o => o);
        return new ObjectConverter<T>(ic);
    }

    private static registerFactory(typeName: string, factory: FactoryType) {
        this.factoryMap.set(typeName, factory);
    }

    private static registerByFunc<T>(typeName: string, parser: ParserFunc<T>, serializer: SerialFunc<T>) {
        this.registerFactory(typeName, () => new InternalConverter(parser, serializer));
    }

    static {
        this.registerByFunc(
            'ListNode', 
            LinkedListFactory.create, 
            LinkedListFactory.dump
        );
        this.registerByFunc(
            'ListNode[]',
            (ls: number[][]) => ls.map((e: number[]) => LinkedListFactory.create(e)!),
            (ts: ListNode[]) => ts.map(e => LinkedListFactory.dump(e)),
        );
        this.registerByFunc(
            'NestedInteger',
            (d: any) => NestedIntegerFactory.parse(d),
            (n: NestedInteger) => NestedIntegerFactory.serialize(n)
        );
        this.registerByFunc(
            'NestedInteger[]',
            (ds: any[]) => NestedIntegerFactory.parseMulti(ds),
            (ns: NestedInteger[]) => NestedIntegerFactory.serializeMulti(ns)
        );
    }

}
